From 286e412967becc7fa48ce0601181e9009695a4f0 Mon Sep 17 00:00:00 2001
From: Quanyang Wang <quanyang.wang@windriver.com>
Date: Mon, 28 Sep 2015 15:00:58 +0800
Subject: [PATCH 987/987] iommu/fsl: remove __initdata annotation of variable
 fsl_of_pamu_driver

In upstream commit 0f1fb99b62ce ("iommu/fsl: Fix section mismatch"), annotate
the variable fsl_of_pamu_driver with __initdata. This means that fsl_of_pamu_driver
will be freed when boot finish, and struct device_driver driver as a struct
member will be freed too. So when call the routine "uevent_store" , "uevent_store"
will run kobject_uevent(&drv->p->kobj, action), but "drv" which embeded in variable
fsl_of_pamu_driver has been freed, so it will occur calltrace as below:

Unable to handle kernel paging request for data at address 0x40feffe0f8c90020^M
Faulting instruction address: 0xc0000000002f349c
Oops: Kernel access of bad area, sig: 11 [#1]
SMP NR_CPUS=24 CoreNet Generic
Modules linked in:
CPU: 3 PID: 1904 Comm: udevadm Not tainted 4.1.6 #3
task: c000000072dd0ac0 ti: c00000007648c000 task.ti: c00000007648c000
NIP: c0000000002f349c LR: c000000000397db8 CTR: 0000000000000000
REGS: c00000007648f700 TRAP: 0300   Not tainted  (4.1.6)
MSR: 0000000080029000 <CE,EE,ME>  CR: 24000882  XER: 20000000
DEAR: 40feffe0f8c90020 ESR: 0000000000000000 SOFTE: 1
GPR00: c000000000397db8 c00000007648f980 c000000000b3c000 40feffe0f8c90000
GPR04: 0000000000000000 0000000000000000 0000000000000003 0000000000000000
GPR08: c00000007256b078 40feffe0f8c90000 0000000000000000 c000000000737718
GPR12: 0000000024000882 c00000000fff6440 000000003a59b170 000000003a59b140
GPR16: 000000003a59b1a0 000000003a57b4f8 000000003a575b28 000000003a59ab64
GPR20: 000000003a59ab60 000000003a57b550 000000003a57b520 000000003a57b5f0
GPR24: 000000003a597040 000000003a577340 c00000007648fe00 c0000000760e2c18
GPR28: fffffffffffffff2 0000000000000003 c000000000a2fd40 0000000000000003
NIP [c0000000002f349c] .kobject_uevent_env+0x64/0x608
LR [c000000000397db8] .uevent_store+0x48/0x68
Call Trace:
[c00000007648f980] [c00000007648fa10] 0xc00000007648fa10 (unreliable)
[c00000007648fa70] [c000000000397db8] .uevent_store+0x48/0x68
[c00000007648fb00] [c000000000397004] .drv_attr_store+0x38/0x54
[c00000007648fb70] [c0000000001a30fc] .sysfs_kf_write+0x5c/0x84
[c00000007648fbe0] [c0000000001a2248] .kernfs_fop_write+0x16c/0x1c4
[c00000007648fc80] [c000000000123994] .__vfs_write+0x30/0x5c
[c00000007648fcf0] [c000000000124330] .vfs_write+0xac/0x1dc
[c00000007648fd90] [c000000000124fb8] .SyS_write+0x58/0xe8
[c00000007648fe30] [c000000000000598] system_call+0x38/0xc4
Instruction dump:
fb41ffd0 fb61ffd8 fb81ffe0 fba1ffe8 fbc1fff0 fbe1fff8 91810008 f821ff11
48000010 e9290018 2fa90000 419e0184 <ebe90020> 2fbf0000 419effec e9230038
---[ end trace c6ea7e28d28c0742 ]---

Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 drivers/iommu/fsl_pamu.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/iommu/fsl_pamu.c b/drivers/iommu/fsl_pamu.c
index 2fea2b48..c26e8c7 100644
--- a/drivers/iommu/fsl_pamu.c
+++ b/drivers/iommu/fsl_pamu.c
@@ -1416,7 +1416,7 @@ error:
 	return ret;
 }
 
-static struct platform_driver fsl_of_pamu_driver __initdata = {
+static struct platform_driver fsl_of_pamu_driver = {
 	.driver = {
 		.name = "fsl-of-pamu",
 	},
-- 
1.9.1

